relayer/crates/starknet-chain-components/src/types/cosmos/client_state.rs

```diff
@@ -101,3 +109,244 @@ impl From<CometClientState> for Any {
         IbcCometClientState::from(client_state).into()
     }
 }
+
+pub struct EncodeLeafOp;
+
+#[cgp_provider(MutEncoderComponent)]
+impl<Encoding, Strategy> MutEncoder<Encoding, Strategy, LeafOp> for EncodeLeafOp
+where
+    Encoding: CanEncodeMut<Strategy, Product![u32, u32, u32, u32, Vec<u8>]>,
+{
+    fn encode_mut(
+        encoding: &Encoding,
+        proof_specs: &LeafOp,
+        buffer: &mut Encoding::EncodeBuffer,
+    ) -> Result<(), Encoding::Error> {
+        let LeafOp {
+            hash,
+            prehash_key,
+            prehash_value,
+            length,
+            prefix,
+        } = serde_json::to_value(proof_specs)
+            .and_then(serde_json::from_value)
+            .unwrap();
+        //.map_err(|_| Encoding::raise_error("invalid connection end"))?;
+        let hash = hash as u32;
+        let prehash_key = prehash_key as u32;
+        let prehash_value = prehash_value as u32;
+        let length = length as u32;
+
+        encoding.encode_mut(
+            &product![hash, prehash_key, prehash_value, length, prefix,],
+            buffer,
+        )?;
+        Ok(())
+    }
+}
+
+#[cgp_provider(MutDecoderComponent)]
+impl<Encoding, Strategy> MutDecoder<Encoding, Strategy, LeafOp> for EncodeLeafOp
+where
+    Encoding: CanDecodeMut<Strategy, Product![u32, u32, u32, u32, Vec<u8>]>
+        + CanRaiseAsyncError<&'static str>,
+{
+    fn decode_mut<'a>(
+        encoding: &Encoding,
+        buffer: &mut Encoding::DecodeBuffer<'a>,
+    ) -> Result<LeafOp, Encoding::Error> {
+        let product![hash, prehash_key, prehash_value, length, prefix] =
+            encoding.decode_mut(buffer)?;
+        Ok(LeafOp {
+            hash: hash as i32,
+            prehash_key: prehash_key as i32,
+            prehash_value: prehash_value as i32,
+            length: length as i32,
+            prefix,
+        })
+    }
+}
+
+pub struct EncodeInnerSpec;
+
+#[cgp_provider(MutEncoderComponent)]
+impl<Encoding, Strategy> MutEncoder<Encoding, Strategy, InnerSpec> for EncodeInnerSpec
+where
+    Encoding: CanEncodeMut<Strategy, Product![Vec<u32>, u32, u32, u32, Vec<u8>, u32]>,
+{
+    fn encode_mut(
+        encoding: &Encoding,
+        proof_specs: &InnerSpec,
+        buffer: &mut Encoding::EncodeBuffer,
+    ) -> Result<(), Encoding::Error> {
+        let InnerSpec {
+            child_order,
+            child_size,
+            min_prefix_length,
+            max_prefix_length,
+            empty_child,
+            hash,
+        } = serde_json::to_value(proof_specs)
+            .and_then(serde_json::from_value)
+            .unwrap();
+        //.map_err(|_| Encoding::raise_error("invalid connection end"))?;
```

Comment:
this one

Reply:
Done in 421fe1fb11ff66b6b39b9353b73f52ee40da5859